java/core/lance-jni/src/error.rs

```diff
@@ -15,149 +15,167 @@
 use std::str::Utf8Error;
 
 use arrow_schema::ArrowError;
-use jni::errors::Error as JniError;
+use jni::{errors::Error as JniError, JNIEnv};
+use lance::error::Error as LanceError;
 use serde_json::Error as JsonError;
-use snafu::{Location, Snafu};
 
-/// Java Exception types
-pub enum JavaException {
-    IllegalArgumentException,
-    IOException,
-    RuntimeException,
+pub type JavaResult<T> = std::result::Result<T, JavaError>;
+
+#[derive(Debug)]
+pub enum JavaExceptionClass {
+  IllegalArgumentException,
+  IOException,
+  RuntimeException,
+  UnsupportedOperationException,
+}
+
+impl JavaExceptionClass {
+  pub fn as_str(&self) -> &str {
+      match self {
+          Self::IllegalArgumentException => "java/lang/IllegalArgumentException",
+          Self::IOException => "java/io/IOException",
+          Self::RuntimeException => "java/lang/RuntimeException",
+          Self::UnsupportedOperationException => "java/lang/UnsupportedOperationException"
+      }
+  }
 }
 
-impl JavaException {
-    pub fn as_str(&self) -> &str {
-        match self {
-            Self::IllegalArgumentException => "java/lang/IllegalArgumentException",
-            Self::IOException => "java/io/IOException",
-            Self::RuntimeException => "java/lang/RuntimeException",
-        }
-    }
+#[derive(Debug)]
+pub struct JavaError {
+  message: String,
```

Comment:
as @wjones127 pointed out in #2316, to get more precise error info, it might be better to put a BoxedError here. like
```rust
#[derive(Debug)]
pub struct JavaError {
    source: BoxedError,
    java_class: JavaExceptionClass,
}
```
BoxedError is defined here:
[BoxedError](https://github.com/lancedb/lance/blob/af926c6d1dd8554851b8cfa616db66f0888772ec/rust/lance-core/src/error.rs#L7)

I haven't read much of the Lance codebase, correct me if I'm wrong

Reply:
@broccoliSpicy Good to know about the BoxedError and its benefits! Thanks for the knowledge transfer!
For java side, this JavaError is a last mile error to the user, which translates the LanceError into Java Error with string message. 
The information in the BoxedError will be included in this String message
e.g. In lance Error, 
    #[snafu(display("Dataset at path {path} was not found: {source}, {location}"))]
    DatasetNotFound {
        path: String,
        source: BoxedError,
        location: Location,
    },
on java side we directly call LanceError.toString() to get the message with source information.

Reply:
@LuQQiu glad to know how error message is transfered from lance to jave, thanks!